src/hip_hcc.cpp

```diff
@@ -2521,4 +2521,19 @@ namespace hip_impl {
             std::terminate();
         #endif
     }
+
+    std::mutex executables_cache_mutex;
+
+    void executables_cache(
```

Comment:
Since the lock is have moved outside of this function, I think it would be simpler and efficient to just return a ref to the std::vector<hsa_executable_t>?  That way, we don't need to make a new copy on read and we won't actually need a write operation.